### ISSUE TITLE
support fp16

### ISSUE THREAD

**Issue by hsmyy** [uber/horovod#278]
does horovod want to support float16 communication?

**Comment by alsrgv** [comment]
@hsmyy, can you provide more context - are you interested in using `float16` in broadcast or allreduce operations?  Are you planning to use [mixed precision](https://devblogs.nvidia.com/mixed-precision-training-deep-neural-networks/) training?

I can push out a branch with FP16 support if you'd like to try and report whether you see the same convergence as with FP32 gradient averaging.

**Comment by hsmyy** [comment]
Hi, @alsrgv , as you said, I want to try mixed precision training using TF+horovod. If you can push out the FP16 branch, I will try it and give you some feedback about the precision and preformance. Thank you very much!

**Comment by alsrgv** [comment]
@hsmyy, sounds good!  Will do that next week.

**Comment by SrivastavaKshitij** [comment]
Hi guys. Any update on fp16 support ?

**Comment by alsrgv** [comment]
Hi @hsmyy,

I've pushed two branches with FP16 support:
1. https://github.com/uber/horovod/tree/fp16 - SUM, then divide by # of workers
2. https://github.com/uber/horovod/tree/fp16_divide_before_sum - divide by # of workers, then SUM.

I'd love if you try both.  The first addresses vanishing gradient problem, while the second addresses gradient overflow problem.

You can install directly from git using:
```
$ <flags> pip install -v --no-cache-dir git+https://github.com/uber/horovod@<branch>
```

Sample code:
```python
import tensorflow as tf
import horovod.tensorflow as hvd

hvd.init()

config = tf.ConfigProto()
config.gpu_options.allow_growth = True
config.gpu_options.visible_device_list = str(hvd.local_rank())
config.log_device_placement = True
sess = tf.Session(config=config)

with tf.device('/gpu:0'):
    c = tf.placeholder(dtype=tf.float16)
    r = hvd.allreduce(c)
    print(sess.run(r, feed_dict={c: hvd.local_rank()}))
```

**NOTE**. CPU version of allreduce is not implemented yet, so if your op is placed on CPU you will get an error like this:
```
[asergeev-C02S638GG8WM:45834] *** An error occurred in MPI_Allreduce: the reduction operation MPI_SUM is not defined for non-intrinsic datatypes
[asergeev-C02S638GG8WM:45834] *** reported by process [3302162433,6880949194907975680]
[asergeev-C02S638GG8WM:45834] *** on communicator MPI_COMM_WORLD
[asergeev-C02S638GG8WM:45834] *** MPI_ERR_OP: invalid reduce operation
[asergeev-C02S638GG8WM:45834] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
[asergeev-C02S638GG8WM:45834] ***    and potentially your MPI job)
```

**Comment by hsmyy** [comment]
Hi @alsrgv , Thank you very much, I will try both branches and give you some feedback!

**Comment by alsrgv** [comment]
@hsmyy, do you have any findings so far?

**Comment by roastduck** [comment]
I tried to run [tensorflow benchmark](https://github.com/tensorflow/benchmarks) and cast fp32 to fp16 before allreduce. In my configuration, sum-before-divide may lead to Segmentation Fault, while divde-before-sum works well.

**Comment by alsrgv** [comment]
@roastduck, do you see a performance improvement with fp16, and/or any difference in convergence with divide-before-sum?  Also, can you share output of:
```
$ gdb python /path/to/core
# thread apply all bt
```

**Comment by roastduck** [comment]
@alsrgv,
1. fp16 nearly doubles the performance of VGG16 model.
2. I only trained it for a short period of time (around 8h). I didn't wait until it to converge.
3. The `fp16` branch (sum before divide) won't crash immediately. It crash after training for several hours. I'm sorry that I cannot provide the dump from gdb, for some reasons.

**Comment by alsrgv** [comment]
@roastduck, would be great if you can capture gdb trace if it happens again.  NaN should not cause a segfault.  Did you observe whether either of branches converge satisfactorily?

**Comment by alsrgv** [comment]
@hsmyy, any luck with your experiments?

**Comment by jeffhammond** [comment]
A few comments:
- I proposed to add this to the MPI standard: https://github.com/mpi-forum/mpi-issues/issues/65.
- There will be no performance benefit to using float16 versus float32 for small messages, because the initiation cost of moving data is higher than the bandwidth costs.  On HPC networks like Blue Gene/Q, the timings for `MPI_Allreduce` are flat up to at least 512 bytes.  I suspect that Ethernet has a much higher initiation cost.
- The truncation error from summing float16 data will be substantial, so I would recommend using float32 for accumulation even if it took longer.
- The built-in datatypes use optimized implementations of collectives that are usually far more efficient than what users can achieve with user-defined collectives (Section V of https://github.com/jeffhammond/BigMPI-paper/blob/master/exampi14_resubmission_3.pdf has some measurements on this).  I suspect the performance benefit of using the optimized implementations outweighs any efficiency gain associated with float16 over float32.  While float16 will move half as many bytes, if the user doesn't exploit topology and shared-memory, they will move quite a few more bytes than a properly optimized implementation of float32.
- If floating-point computation is the bottleneck, which is unlikely, then at least on Intel processors, there is no benefit to float16, because the hardware does the compute using float32 units after efficient conversion from float16 storage (https://software.intel.com/en-us/articles/performance-benefits-of-half-precision-floats).

I recommend that you reach out the developers of the MPI implementation(s) you use and request they implement float16 collectives internally using the same optimized code as is available for float32.  You should also request that this implementation use compensated summation or float32 accumulators to avoid truncation error.

One way to roll your own optimized implementation would be to modify [libNBC](http://htor.inf.ethz.ch/research/nbcoll/libnbc/), which is written by one of the best MPI developers on earth and should be using good algorithms.  It's the basis for nonblocking collectives in Open-MPI.

**Comment by alsrgv** [comment]
@jeffhammond, thanks for the feedback and the pointers!

We purposely did not implement fp16 CPU reduction in this first phase because, as you mentioned, it lacks many of performance benefits.  We will look into doing so before the official release.

I am not certain that aggregation in fp32 is necessary for the deep learning domain.  We will run convergence experiments with pure fp16 types on GPU to determine if there's any loss of accuracy, and based on those results decide how to proceed.

**Comment by jeffhammond** [comment]
@alsrgv Indeed, fp16 won't lose much if you have numbers with similar exponents.  However, if you ever lose bits in the summation, there is the potential for non-determinism from MPI.  While I know of no MPI implementation that uses a nondeterministic algorithm for reductions, the MPI standard permits it, and the performance benefits may be significant enough to motivate implementing it in the future.

**Comment by alsrgv** [comment]
Closing this. CPU FP16 with Intel intrinsics has been merged into master and will be released in Horovod 0.15.1.
